File: apps/daemon/internal/logger/redactor.go
// Package logger's redactor is a thin shim over the canonical redaction
// package (`apps/daemon/internal/redaction/`, hp-je1p). Non-logger
// consumers (audit log, WS event fan-out, adapter capture) import that
// package directly; the logger uses it via this shim to keep the hp-lxs
// API stable.
package logger

import (
	"github.com/hoopoe-cockpit/hoopoe/apps/daemon/internal/redaction"
)

// Redactor is the canonical daemon redactor. Re-exported for hp-lxs
// back-compat.
type Redactor = redaction.Redactor

// RedactionEvent is a re-export of redaction.TraceEvent — the canonical
// trace shape consumed across audit / events / logger / adapter capture.
type RedactionEvent = redaction.TraceEvent

// NewRedactor returns a Redactor with the canonical Hoopoe patterns.
func NewRedactor() *Redactor { return redaction.NewDefault() }

// RedactEntry runs the redactor over an entry's msg + every fields entry
// in place; returns the trace events that fired. Lives in the logger
// package because it's specific to the Entry envelope shape.
func RedactEntry(r *Redactor, e *Entry) []RedactionEvent {
	if r == nil || e == nil {
		return nil
	}
	var events []RedactionEvent
	if e.Msg != "" {
		out, msgEvents := r.RedactText(redaction.SurfaceLogger, "msg", e.Msg)
		e.Msg = out
		events = append(events, msgEvents...)
	}
	if len(e.Fields) > 0 {
		out := make(map[string]any, len(e.Fields))
		for k, v := range e.Fields {
			redacted, fieldEvents := r.RedactValue(redaction.SurfaceLogger, "fields."+k, v)
			out[k] = redacted
			events = append(events, fieldEvents...)
		}
		e.Fields = out
	}
	return events
}

File: apps/desktop/src/shared/logger/redactor.ts
// hp-je1p extracted the redactor primitive into `apps/desktop/src/shared/
// redact/`. This file remains as a thin shim so existing call sites
// (notably hp-lxs tests + the in-process logger redaction path) keep
// compiling without source changes.

import type { LogEntry } from "./types.ts";
import {
  Redactor,
  SurfaceLogger,
  type RedactValue,
  type TraceEvent,
} from "../redact/index.ts";

export { Redactor };
/** Re-exported for hp-lxs back-compat. */
export type RedactionEvent = TraceEvent;

/** redactLogEntry: scrubs the given LogEntry's msg + every value in
 *  fields. Returns the redacted entry + the events that fired. Envelope
 *  columns (component / jobId / ...) are NOT redacted. */
export function redactLogEntry(
  redactor: Redactor,
  entry: LogEntry,
): { entry: LogEntry; events: TraceEvent[] } {
  const events: TraceEvent[] = [];
  let msg = entry.msg;
  if (msg) {
    const { redacted, events: msgEvents } = redactor.redactText(SurfaceLogger, "msg", msg);
    msg = redacted;
    if (msgEvents.length) events.push(...msgEvents);
  }
  let fields = entry.fields;
  if (fields) {
    const out: Record<string, RedactValue> = {};
    for (const [k, v] of Object.entries(fields)) {
      const { redacted, events: fieldEvents } = redactor.redactValue(
        SurfaceLogger,
        `fields.${k}`,
        v as RedactValue,
      );
      out[k] = redacted as RedactValue;
      if (fieldEvents.length) events.push(...fieldEvents);
    }
    fields = out as LogEntry["fields"];
  }
  const next: LogEntry = { ...entry, msg };
  if (fields !== undefined) {
    Object.assign(next, { fields });
  }
  return { entry: next, events };
}
